src/main/scala/org/apache/spark/mllib/tfocs/Mode.scala

```diff
@@ -0,0 +1,25 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.tfocs
+
+/**
+ * Evaluation mode.
+ * @param f whether to compute function value
+ * @param g whether to compute gradient
+ */
+case class Mode(f: Boolean, g: Boolean)
```

Comment:
I was thinking about switching to Int as well, but then I thought if the user wants both f and g (Mode(true,true)), then g will be accessed twice so it should be cached (following the simple rule of caching if accessed twice). Otherwise the smooth function implementation doesn’t need to cache it.

More detailed description: Let’s say f and g are computed using a shared dependency (call it x). (In some cases like the linear regression implementation g and x are the same). The question is whether x should be cached. If we follow the rule “cache x if it will be used at least twice” then if Mode is (true, true) x will be needed twice so we cache it. If Mode is (true, false) then only f is computed and x is an intermediate RDD that does not need to be cached. If Mode is false, true, then x is an intermediate RDD to compute g. And g is going to be returned, so the user can decide if g should be cached or not after it is returned.

Here’s what I implemented:
https://github.com/staple/spark-agd/pull/2/files#diff-b3a1cbdf9bb73db6a61ca6314d1e7e8dR42

Let me know if you have any more thoughts though.